holoviews/streams.py

```diff
@@ -0,0 +1,220 @@
+
+
+
+class Preprocessor(param.Parameterized):
+    """
+    A Preprocessor is a callable that takes a dictionary as an argument
+    and returns a dictionary. Where possible, Preprocessors should have
+    valid reprs that can be evaluated.
+
+    Preprocessors are used to set the value of a stream based on the
+    parameter values. They may be used for debugging purposes or to
+    remap or repack parameter values before they are passed onto to the
+    subscribers.
+    """
+
+    def __call__(self, params):
+        return params
+
+
+
+class Rename(Preprocessor):
+    """
+    A preprocessor used to rename parameter values.
+    """
+
+    mapping = param.Dict(default={}, doc="""
+      The mapping from the parameter names to the designated names""")
+
+    def __init__(self, **mapping):
+        super(Rename, self).__init__(mapping=mapping)
+
+    def __call__(self, params):
+        return {self.mapping.get(k,k):v for (k,v) in params.items()}
+
+    def __repr__(self):
+        keywords = ','.join('%s=%r' % (k,v) for (k,v) in sorted(self.mapping.items()))
+        return 'Rename(%s)' % keywords
+
+
+
+class Group(Preprocessor):
+    """
+    A preprocessor that keeps the parameter dictionary together,
+    supplying it as a value associated with the given key.
+    """
+
+    def __init__(self, key):
+        super(Group, self).__init__(key=key)
+
+    def __call__(self, params):
+        return {self.key:params}
+
+    def __repr__(self):
+        return 'Group(%r)' % self.key
+
+
+
+class Stream(param.Parameterized):
+    """
+    A Stream is simply a parameterized object with parameters that
+    change over time in response to update events. Parameters are
+    updated via the update method.
+
+    Streams may have one or more subscribers which are callables passed
+    the parameter dictionary when the trigger classmethod is called.
+    """
+
+    # Mapping from uuid to stream instance
+    registry = OrderedDict()
+
+    @classmethod
+    def trigger(cls, streams):
+        """
+        Given a list of streams, collect all the stream parameters into
+        a dictionary and pass it to the union set of subscribers.
+
+        Passing multiple streams at once to trigger can be useful when a
+        subscriber may be set multiple times across streams but only
+        needs to be called once.
+        """
+        # Union of stream values
+        items = [stream.value.items() for stream in streams]
+        union = [kv for kvs in items for kv in kvs]
+        klist = [k for k,_ in union]
+        clashes = set([k for k in klist if klist.count(k) > 1])
+        if clashes:
+            param.main.warning('Parameter name clashes for keys: %r' % clashes)
+
+        # Currently building a simple set of subscribers
+        groups = [stream.subscribers + stream._hidden_subscribers for stream in streams]
+        subscribers = set(s for subscribers in groups for s in subscribers)
+        for subscriber in subscribers:
```

Comment:
Hmm, shouldn't some order be maintained, so that the subscribers defined on an individual stream are at least executed in sequence and the _hidden_subscribers are executed after all others?